src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs

```diff
@@ -24,34 +24,41 @@ public class ArgumentCompleterAttribute : Attribute
     {
         /// <summary/>
         [SuppressMessage("Microsoft.Naming", "CA1721:PropertyNamesShouldNotMatchGetMethods")]
-        public Type Type { get; private set; }
+        public Type Type { get; }
 
         /// <summary/>
-        public ScriptBlock ScriptBlock { get; private set; }
+        public ScriptBlock ScriptBlock { get; }
 
         /// <param name="type">The type must implement <see cref="IArgumentCompleter"/> and have a default constructor.</param>
         public ArgumentCompleterAttribute(Type type)
         {
             if (type == null || (type.GetInterfaces().All(t => t != typeof(IArgumentCompleter))))
             {
-                throw PSTraceSource.NewArgumentException("type");
+                throw PSTraceSource.NewArgumentException(nameof(type));
             }
 
             Type = type;
         }
 
         /// <summary>
-        /// This constructor is used primarily via PowerShell scripts.
+        /// Initializes a new instance of the <see cref="ArgumentCompleterAttribute"/> class.
+        /// This constructor is used by derived attributes implementing IArgumentCompleterFactory.
         /// </summary>
-        /// <param name="scriptBlock"></param>
-        public ArgumentCompleterAttribute(ScriptBlock scriptBlock)
+        protected ArgumentCompleterAttribute()
         {
-            if (scriptBlock == null)
+            if (GetType().GetInterfaces().All(t => t != typeof(IArgumentCompleterFactory)))
             {
-                throw PSTraceSource.NewArgumentNullException("scriptBlock");
+                throw PSTraceSource.NewInvalidOperationException();
             }
+        }
```

Comment:
I don't see where it is used.

Reply:
It isn't used yet. This is infrastructure intended to be used by others

Reply:
Could you please add more info how should it be used? Maybe the doc comment should be better.

Reply:
Good point. Adding an example usage in the doc comments